Mathlib/Order/Category/Frm.lean

```diff
@@ -22,63 +22,138 @@ universe u
 
 namespace Frm
 
-instance : CoeSort Frm Type* :=
-  Bundled.coeSort
+instance : CoeSort Frm (Type _) :=
+  ⟨Frm.carrier⟩
+
+attribute [coe] Frm.carrier
+
+/-- Construct a bundled `Frm` from the underlying type and typeclass. -/
+abbrev of (X : Type*) [Frame X] : Frm := ⟨X⟩
+
+/-- The type of morphisms in `Frm R`. -/
+@[ext]
+structure Hom (X Y : Frm.{u}) where
+  private mk ::
+  /-- The underlying `FrameHom`. -/
+  hom' : FrameHom X Y
+
+instance : Category Frm.{u} where
+  Hom X Y := Hom X Y
+  id X := ⟨FrameHom.id X⟩
+  comp f g := ⟨g.hom'.comp f.hom'⟩
+
+instance : ConcreteCategory Frm (FrameHom · ·) where
+  hom := Hom.hom'
+  ofHom := Hom.mk
+
+/-- Turn a morphism in `Frm` back into a `FrameHom`. -/
+abbrev Hom.hom {X Y : Frm.{u}} (f : Hom X Y) :=
+  ConcreteCategory.hom (C := Frm) f
+
+/-- Typecheck a `FrameHom` as a morphism in `Frm`. -/
+abbrev ofHom {X Y : Type u} [Frame X] [Frame Y] (f : FrameHom X Y) : of X ⟶ of Y :=
+  ConcreteCategory.ofHom (C := Frm) f
 
-instance (X : Frm) : Frame X :=
-  X.str
+variable {R} in
+/-- Use the `ConcreteCategory.hom` projection for `@[simps]` lemmas. -/
+def Hom.Simps.hom (X Y : Frm.{u}) (f : Hom X Y) :=
+  f.hom
 
-/-- Construct a bundled `Frm` from a `Frame`. -/
-def of (α : Type*) [Frame α] : Frm :=
-  Bundled.of α
+initialize_simps_projections Hom (hom' → hom)
+
+/-!
+The results below duplicate the `ConcreteCategory` simp lemmas, but we can keep them for `dsimp`.
+-/
 
 @[simp]
-theorem coe_of (α : Type*) [Frame α] : ↥(of α) = α := rfl
+lemma coe_id {X : Frm} : (𝟙 X : X → X) = id := rfl
 
-instance : Inhabited Frm :=
-  ⟨of PUnit⟩
+@[simp]
+lemma coe_comp {X Y Z : Frm} {f : X ⟶ Y} {g : Y ⟶ Z} : (f ≫ g : X → Z) = g ∘ f := rfl
+
+@[simp]
+lemma forget_map {X Y : Frm} (f : X ⟶ Y) :
+    (forget Frm).map f = f := rfl
 
-/-- An abbreviation of `FrameHom` that assumes `Frame` instead of the weaker `CompleteLattice`.
-Necessary for the category theory machinery. -/
-abbrev Hom (α β : Type*) [Frame α] [Frame β] : Type _ :=
-  FrameHom α β
+@[ext]
+lemma ext {X Y : Frm} {f g : X ⟶ Y} (w : ∀ x : X, f x = g x) : f = g :=
+  ConcreteCategory.hom_ext _ _ w
 
-instance bundledHom : BundledHom Hom where
-  toFun {α β} _ _ := ((↑) : FrameHom α β → α → β)
-  id {α} _ := FrameHom.id α
-  comp _ _ _ := FrameHom.comp
-  hom_ext _ _ := DFunLike.coe_injective
+-- This is not `simp` to avoid rewriting in types of terms.
+theorem coe_of (X : Type u) [Frame X] : (Frm.of X : Type u) = X := rfl
+
+@[simp]
+lemma hom_id {X : Frm} : (𝟙 X : X ⟶ X).hom = FrameHom.id _ := rfl
 
--- Porting note: Originally `deriving instance LargeCategory, HasForget for Frm`
--- see https://github.com/leanprover-community/mathlib4/issues/5020
-deriving instance LargeCategory, Category for Frm
+/- Provided for rewriting. -/
+lemma id_apply (X : Frm) (x : X) :
+    (𝟙 X : X ⟶ X) x = x := by simp
 
-instance : HasForget Frm := by
-  unfold Frm
-  infer_instance
+@[simp]
+lemma hom_comp {X Y Z : Frm} (f : X ⟶ Y) (g : Y ⟶ Z) :
```

Comment:
Don't you feel like there should be automation for this now?

Reply:
Mhm, it would be helpful although I don't quite know what form that would take. It's hard to state this inside the type theory itself, since for each base morphism (that is `FrameHom`, not `Frm.Hom`) the type of `comp` can take a different shape. Maybe a metaprogram generating all the boilerplate?